### ISSUE TITLE
RFC: `sub_select` should have a nullable result type by default

### ISSUE THREAD

**Issue by parsonsmatt** [bitemyapp/esqueleto#151]
tl;dr: `sub_select` should have the type `sub_select :: SqlQuery (SqlExpr (Value a)) -> SqlExpr (Value (Maybe a))`.

We use `sub_select` extensively in the work codebase, and it is a common source of runtime errors. The general pattern is this:

```haskell
bad :: SqlPersistM [(Entity Foo, XPropType)]
bad = 
  select $
  from $ \foo -> do 
  where_ (foo ^. FooProperty ==. val 3)
  pure (foo, sub_select (from_ \x -> where_ (p x) >> limit 1 >> pure (x ^. XProp))
```

`sub_select` has the type `sub_select :: SqlQuery (SqlExpr (Value a)) -> SqlExpr (Value a)`. Because the result type is `Value a`, the relevant marshalling code assumes that it cannot be `NULL`. However, it totally can be `NULL`, for a huge amount of the potential uses - possibly even the majority or default case.

A problem is that it is going to be quite difficult to statically determine whether or not it will return null. `sub_select $ from $ \x -> where_ (p x) >> limit 1 >> pure (x ^. XProp)` will obviously return `NULL` if the query returns 0 results, and `persistent` will fail to parse a `Nothing`. The work-around is to use `just`.

However, `sub_select $ from $ \x -> where_ (p x) >> pure (sum (x ^. XProp))` will return `0` for zero rows, so it doesn't *need* to be `Maybe`. Likewise, `sub_select $ from $ \x -> where_ (p x) >> pure countRows` will also return a non-null result.

I propose that `sub_select` be made nullable by default, along with another function that can dismiss the `Maybe` when the user knows exactly what they're doing and it'll be fine.

**Comment by bitemyapp** [comment]
It'll break a lot of existing code, probably, but this sounds like the right thing to do to me.

**Comment by parsonsmatt** [comment]
I've implemented #153 in the work codebase. It's made a *lot* of code safer and unearthed a few potential bugs, while also reducing boilerplate. 

Given the deprecation message with the helpful warning, I think it shouldn't be too terribly onerous of a migration path and upgrade.
